lsm: fix issue with level0 compactions and overlapping files

rockwotj opened the pull request:
Now there is a very subtle issue in compaction of level0 files in that
if we expand the set of files due to an overlap, that file itself could
have caused an additional overlap, so we need to reset the search and
include the expanded space. This is due to level 0 allowing overlap of
the key space. When we did this reset, we were incrementing the counter
after we set i to 0. The proper fix, is to use the exact logic from
leveldb and only increment the counter after we select the file so the
resetting of `i` works.

This was found while writing a bunch of tests for the compaction picking
logic. So yay! for tests :)

Now this would never effect actual correctness of the LSM tree, but could
mean that compaction could in theory always skip compacting some L0 file.

LevelDB link:
https://github.com/google/leveldb/blob/ac691084fdc5546421a55b25e7653d450e5a25fb/db/version_set.cc#L498

Additionally there were some bugs around tombstones where some keys could be made visible again that were deleted. Originally I had hoped we didn't need some of the messy things leveldb does with user keys, so I had omitted them. Turns out those are actually needed (and there is a test for it now!), but thankfully I was able to find a solution that is a little less clunky IMO

## Backports Required

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x

## Release Notes

* none

rockwotj committed:
lsm: fix issue with level0 compactions and overlapping files

Now there is a very subtle issue in compaction of level0 files in that
if we expand the set of files due to an overlap, that file itself could
have caused an additional overlap, so we need to reset the search and
include the expanded space. This is due to level 0 allowing overlap of
the key space. When we did this reset, we were incrementing the counter
after we set i to 0. The proper fix, is to use the exact logic from
leveldb and only increment the counter after we select the file so the
resetting of `i` works.

This was found while writing a bunch of tests for the compaction picking
logic. So yay! for tests :)

LevelDB link:
https://github.com/google/leveldb/blob/ac691084fdc5546421a55b25e7653d450e5a25fb/db/version_set.cc#L498

rockwotj committed:
lsm: fix get_range bug

We copy/pasted this line and messed up the conditional to get the max
value. This adds tests for these functions and fixes the compaction
tests that were wrong because of it.

rockwotj committed:
lsm: add tests for compaction pointers

The compaction pointer is important to ensure we rotate properly through
the entire key space when computing compactions. Without it, we could
starve part of the key space from compactions.

We just add tests for this and restore a helpful comment from the
leveldb source about why we immediately update the compaction pointer in
addition to applying it to the edit.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--243e84a3-54ba-40f8-b169-64049ad360e7-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/78019">build#78019</a></summary>


| test_class              | test_method             | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                              |
|:------------------------|:------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| MountUnmountIcebergTest | test_simple_remount     | {"cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/78019#019b2aa1-d081-4721-bad5-ba4db2b790ad | FLAKY         | 9/11     | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.1955, p0=0.8864, reject_threshold=0.0100. adj_baseline=0.4792, p1=0.0150, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount |
| ScalingUpTest           | test_fast_node_addition | null                      | integration | https://buildkite.com/redpanda/redpanda/builds/78019#019b2aa1-d081-4721-bad5-ba4db2b790ad | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0379, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1093, p1=0.3142, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ScalingUpTest&test_method=test_fast_node_addition       |
</details>
<!--/body-->

rockwotj committed:
lsm: add test exposing issue with user key comparison

See comments in the test, there will be a series of fixes for this
general problem following this.

rockwotj committed:
lsm: introduce strongly typed user_key

We want to use user keys in some places for comparison, so let's
introduce a strong type to make that more obvious.

rockwotj committed:
lsm: compare file boundaries using user keys in compaction

See the added test, this is required for compaction of tombstones,
otherwise we could incorrectly drop tombstones early, causing values to
be resurface that are deleted.

The fix is to use our new user_key type to enforce we compare by user
keys when comparing file boundaries for determining which files need
to be input to a compaction.
